src/chain.ts

```diff
@@ -14,47 +17,60 @@ const createHttpResponse = (statusCode, message) => {
     ].join('\r\n');
 };
 
-/**
- * @typedef Options
- *
- * @property {ClientRequest} request
- * @property {net.Socket} sourceSocket - a stream where to pipe from
- * @property {Buffer} head - optional, the response buffer attached to CONNECT request
- * @property {*} handlerOpts - handler options that contain upstreamProxyUrlParsed
- * @property {http.Server} server - the server that we will use for logging
- * @property {boolean} isPlain - whether to send HTTP CONNECT response
- */
-
-/**
- * @param {Options} options
- */
-const chain = ({ request, sourceSocket, head, handlerOpts, server, isPlain }) => {
+interface Options {
+    method: string;
+    headers: string[];
+    path?: string;
+}
+
+export interface HandlerOpts {
+    upstreamProxyUrlParsed: URL;
+}
+
+export const chain = (
+    {
+        request,
+        sourceSocket,
+        head,
+        handlerOpts,
+        server,
+        isPlain,
+    }: {
+        request: { url?: string },
+        sourceSocket: net.Socket,
+        head?: Buffer,
+        handlerOpts: HandlerOpts,
+        server: EventEmitter & { log: (...args: any[]) => void; },
+        isPlain: boolean,
+    },
+): void => {
     if (head && head.length > 0) {
         throw new Error(`Unexpected data on CONNECT: ${head.length} bytes`);
     }
 
-    const { proxyChainId } = sourceSocket;
+    const { proxyChainId } = sourceSocket as unknown as { proxyChainId: unknown };
 
     const { upstreamProxyUrlParsed: proxy } = handlerOpts;
 
-    const options = {
+    const options: Options = {
         method: 'CONNECT',
         path: request.url,
         headers: [
             'host',
-            request.url,
+            request.url!,
         ],
     };
 
     if (proxy.username || proxy.password) {
         options.headers.push('proxy-authorization', getBasic(proxy));
     }
 
-    const client = http.request(proxy.origin, options);
+    const client = http.request(proxy.origin, options as unknown as http.ClientRequestArgs);
```

Comment:
The issue is that `@types/node` doesn't support an array for `headers`. We're using an array because it's easier to mutate it and it's much faster than a plain object.